### ISSUE TITLE
Performance gets affected when using 'limit' as it applies limit after the table join

### ISSUE THREAD

**Issue by reidharry** [hasura/graphql-engine#5745]
### Summary 

Since 'limit' is getting applied after the table joins, the performance is greatly affected. If the limit for the first table is applied and then the table join happens, it would greatly improve the performance.

#### Sample database schema
<details>
  <summary>Click to expand!</summary>
  
[workmanage.sql.zip](https://github.com/hasura/graphql-engine/files/5186091/workmanage.sql.zip)

</details>

#### GraphQL
<details>
  <summary>Click to expand!</summary>
  
```gql
query MyQuery {
  users(limit: 10, offset: 0) {
    works {
      id
      name
      updated_at
      created_at
    }
    name
    id
    is_active
    updated_at
    username
    created_at
  }
}
```
</details>

#### Generated SQL
<details>
  <summary>Click to expand!</summary>
  
```sql
SELECT
  coalesce(json_agg("root"), '[]') AS "root"
FROM
  (
    SELECT
      row_to_json(
        (
          SELECT
            "_5_e"
          FROM
            (
              SELECT
                "_4_root.ar.root.works"."works" AS "works",
                "_0_root.base"."name" AS "name",
                "_0_root.base"."id" AS "id",
                "_0_root.base"."is_active" AS "is_active",
                "_0_root.base"."updated_at" AS "updated_at",
                "_0_root.base"."username" AS "username",
                "_0_root.base"."created_at" AS "created_at"
            ) AS "_5_e"
        )
      ) AS "root"
    FROM
      (
        SELECT
          *
        FROM
          "public"."users"
        WHERE
          ('true')
      ) AS "_0_root.base"
      LEFT OUTER JOIN LATERAL (
        SELECT
          coalesce(json_agg("works"), '[]') AS "works"
        FROM
          (
            SELECT
              row_to_json(
                (
                  SELECT
                    "_2_e"
                  FROM
                    (
                      SELECT
                        "_1_root.ar.root.works.base"."id" AS "id",
                        "_1_root.ar.root.works.base"."name" AS "name",
                        "_1_root.ar.root.works.base"."updated_at" AS "updated_at",
                        "_1_root.ar.root.works.base"."created_at" AS "created_at"
                    ) AS "_2_e"
                )
              ) AS "works"
            FROM
              (
                SELECT
                  *
                FROM
                  "public"."works"
                WHERE
                  (("_0_root.base"."id") = ("user_id"))
              ) AS "_1_root.ar.root.works.base"
          ) AS "_3_root.ar.root.works"
      ) AS "_4_root.ar.root.works" ON ('true')
    LIMIT
      10 OFFSET ('0') :: integer
  ) AS "_6_root"
```
</details>

**Comment by 0x777** [comment]
Hi @reidharry, the `limit` is currently added at the outer level because the ordering could use columns from relationships. Postgres's query planner should be able to push the limit into the inner table automatically? What does the explain output say? Either way, we'll definitely improve this - `limit` will be added at the right level based on the `order_by` clause.

**Comment by irees** [comment]
I'm also affected by this - queries that are quite fast with the limit placed on the inner query are extremely slow when the limit is on the outer, fully joined product - even when no part of the join is used in the sorting or filtering.

**Comment by Nohac** [comment]
I'm also experiencing this problem. I copied the query from the "analyze" page and moved limit to the "main" select statement, and the query time went from about ~20 seconds to less than a second.

**Comment by meotimdihia** [comment]
This issue was tagged as a priority high but still has not been fixed. I wonder everyone forgot it.

**Comment by natew** [comment]
You'd think something moving onto 2.0 with multi-db support would have the very basics of joining down.

**Comment by paf31** [comment]
When testing this locally with your repro case (thank you for providing that!), I see dramatically better performance on the small data set when adding an index on `works (user_id)`. We are currently looking into code generation changes which will nudge the planner in the right direction in general, but short term, is it possible to try adding the index to see if this can be addressed that way?

**Comment by Nohac** [comment]
I found a nonideal workaround for this by splitting up the pagination into two queries.

The first query that has the `offset` and `limit` only selects the top-level id's and performs no joins, then I pass the id's from the first query to another query that performs all the joins.

This approach ended up being a lot faster then doing it all in one go.

**Comment by paf31** [comment]
@Nohac Thank you for documenting this workaround. Did you also try adding an index to the foreign key responsible for the join in question (and ideally making sure any statistics are freshly computed)?

**Comment by lowercase00** [comment]
Any news? Sub-seconds raw SQL queries in Postgres are taking +20 seconds in Hasura in a 100k row table,  making it completely unusable. It feels this is really high priority, since it's a deal breaker on using Hasura.

**Comment by lowercase00** [comment]
Sorry for sounding repetitive, but this doesn't sound like a "performance" issue in the sense that makes things runs faster. It is not that it is ok to use as is and it would be nice if it was X seconds faster.  As it is, the feature is unusable, so I would say this is actually a bug.

Don't get me wrong, I understand it may not be easy (or fast) to fix, but just found it kind of weird to call it a performance issue.

**Comment by lowercase00** [comment]
Hi, I've seen this is closed with commit [fa152d8](https://github.com/hasura/graphql-engine/commit/fa152d842deb24786d8591a8f347b2f2f78f62b4), it seems to pass the tests, but I couldn't find it in the Changelog, I think it may be the first line on the next release (server: optimize SQL query generation with LIMITs), but this is not market with the issue #, so I'm not really sure. Just was wondering if it's the same issue. Thanks!

**Comment by prisira** [comment]
I think this has been fixed in
https://github.com/hasura/graphql-engine/releases/tag/v2.1.0-beta.1
> server: optimize SQL query generation with LIMITs

Congratulations!
